Tutorial for debugging a failing test using docker (#489)
ros-planning/moveit_tutorials: doc/test_debugging/test_debugging_tutorial.rst

Debugging Tests

> ================

Reviewer: ```suggestion
===============
```

How to debug when a test is failing.

**Note:** *This is not meant as an exhaustive tutorial on software testing, instead this focuses on methods that will help you debug tests in MoveIt or similar ROS projects.*

Getting Started
---------------
If you haven't already done so, make sure you've completed the steps in `Getting Started <../getting_started/getting_started.html>`_.

CI Failures
-----------

Our CI runs in Travis and uses scripts found in the `moveit_ci repo <https://github.com/ros-planning/moveit_ci.git>`.  These tests build and run various tests in various environments.  Often something that works locally won't work in CI in a different environment.  To troubleshoot a failure from CI it is useful to use docker to run in the same environment.

For troubleshooting a specific travis test it is helpful to look at the .travis.yml config file and test output to understand what environment variables are being set in your test.

Launch Docker Environment

> -----------------------

Reviewer: ```suggestion
-------------------------
```

To start docker container for kinetic:

  docker pull moveit/moveit:kinetic-ci
  docker run -it moveit/moveit:kinetic-ci /bin/bash

To start docker container for melodic:

  docker pull moveit/moveit:melodic-ci
  docker run -it moveit/moveit:melodic-ci /bin/bash

Setup Environment

> ----------------

Reviewer: ```suggestion
-----------------
```

Debugging Tests
================

How to debug when a test is failing.

> **Note:** *This is not meant as an exaustive tutorial on software testing, instead this focueses on methods that will help you debug tests in moveit or similar ros projects.*

Reviewer 2: ```suggestion
**Note:** *This is not meant as an exhaustive tutorial on software testing, instead this focuses on methods that will help you debug tests in MoveIt or similar ROS projects.*
```

Getting Started
---------------
If you haven't already done so, make sure you've completed the steps in `Getting Started <../getting_started/getting_started.html>`_.

CI Failures
-----------

> Our CI runs in travis and uses scripts found in the `moveit_ci repo <https://github.com/ros-planning/moveit_ci.git>`.  These test build nad run various tests in various enviroments.  Often something that works locally won't work in CI in a different enviroment.  To troubleshoot a failure from CI it is useful to use docker to run in the same enviroment.

Reviewer 2: ```suggestion
Our CI runs in Travis and uses scripts found in the `moveit_ci repo <https://github.com/ros-planning/moveit_ci.git>`.  These tests build and run various tests in various environments.  Often something that works locally won't work in CI in a different environment.  To troubleshoot a failure from CI it is useful to use docker to run in the same environment.
```

For troubleshooting a specific travis test it is helpful to look at the .travis.yml config file and test output to understand what enviroment variables are being set in your test.

Launch Docker Eviroment
-----------------------

To start docker container for kinetic:

  docker pull moveit/moveit:kinetic-ci
  docker run -it moveit/moveit:kinetic-ci /bin/bash

To start docker container for melodic:

  docker pull moveit/moveit:melodic-ci
  docker run -it moveit/moveit:melodic-ci /bin/bash

Setup Enviroment
----------------

The first thing you should do is update debians and install tools you'll need for debugging.  The update is important because that is what we do in CI.

  apt-get update
  apt-get dist-upgrade
  apt-get install -y python-catkin-tools ssh git gdb valgrind vim

> Then you'll want to set enviroment variables important for your test.  In my example I'm using a kinetic enviroment so I set that here.

Reviewer 2: ```suggestion
Then you'll want to set environment variables important for your test.  In my example, I'm using a kinetic environment so I set that here.
```

>   ROS_DISTRO=kinetic

Reviewer 2: This should be defined within the ROS container by default.

> Lastly, you'll want to create the folder structure for yoru ros enviroment:

Reviewer 2: ```suggestion
Lastly, you'll want to create the folder structure for your ROS environment:
```
